### DISCUSSION TITLE
ESLint installing multiple versions

### DISCUSSION THREAD

**Discussion by erosman** [eslint/eslint#18257] [Q&A]
Subsequent to [Discrepancy in the recommended](https://github.com/eslint/config-inspector/discussions/15), I have uninstalled ESLint and npm and reinstalled them again (_Ubuntu 23.10_).

```
sudo apt-get install npm

npm i eslint@next --save-dev
```

I get 2 sets of files:

- `/home/***/node_modules/@eslint`
- `/home/***/node_modules/eslint`

The structure of the second one also appears to be strange.
`/home/***/node_modules/eslint/node_modules/eslint/node_modules/@eslint`

```
/home/***/node_modules/@eslint/js/package.json -> "version": "9.0.0-rc.0",


/home/***/node_modules/eslint/package.json -> "version": "9.0.0-rc.0",
/home/***/node_modules/eslint/node_modules/eslint/package.json -> "version": "8.57.0",
/home/***/node_modules/eslint/node_modules/eslint/node_modules/@eslint/js/package.json -> "version": "8.57.0",
```

```
npm why -r @eslint/js

@eslint/js@9.0.0-rc.0 dev
node_modules/@eslint/js
  @eslint/js@"9.0.0-rc.0" from eslint@9.0.0-rc.0
  node_modules/eslint
    dev eslint@"^9.0.0-rc.0" from the root project

@eslint/js@8.57.0 dev peer
node_modules/eslint/node_modules/eslint/node_modules/@eslint/js
  @eslint/js@"8.57.0" from eslint@8.57.0
  node_modules/eslint/node_modules/eslint
    peer eslint@"^6.0.0 || ^7.0.0 || >=8.0.0" from @eslint-community/eslint-utils@4.4.0
    node_modules/eslint/node_modules/@eslint-community/eslint-utils
      @eslint-community/eslint-utils@"^4.2.0" from eslint@9.0.0-rc.0
      node_modules/eslint
        dev eslint@"^9.0.0-rc.0" from the root project
      @eslint-community/eslint-utils@"^4.2.0" from eslint@8.57.0
```

Why are there both ESLInt 8.57.0 & 9.0.0-rc.0 present?
Is that how it should be?
If not, how should it be fixed?

**Comment by mdjermanovic** [comment]
It could be that the v8.57.0 packages are installed because `@eslint-community/eslint-utils` has `"eslint": "^6.0.0 || ^7.0.0 || >=8.0.0"` in [`peerDependencies`](https://github.com/eslint-community/eslint-utils/blob/fdab467c95cb6a9fcb98f7fef4ce1a7f7ecb2d5d/package.json#L73C6-L75) and prerelease versions don't satisfy this range.

If you want to make sure that the latest prerelease version of `@eslint/js` is available for your config file (i.e., installed at the top level), it would be better to install it directly:

```
npm i eslint@next @eslint/js@next --save-dev
```

**Reply by erosman** [reply]
Do I have to uninstall them first?

**Reply by mdjermanovic** [reply]
Yeah, it's usually better to remove `node_modules` first.

**Comment by erosman** [comment]
I came across pre-release version issue when trying to install `@stylistic/eslint-plugin-js `as well and had to use `--force`.
See also:
[unable to resolve dependency tree: eslint@9.0.0-rc.0](https://github.com/eslint-stylistic/eslint-stylistic/discussions/304)

**Reply by mdjermanovic** [reply]
`--force` is often necessary when installing prerelease versions as they don't satisfy typical semver ranges.

**Comment by erosman** [comment]
I uninstalled and reinstalled, but npm reinstalls ESLInt 8.57.0 automatically as it is required by other packages that are installed.
I guess I will have to wait for the official release of ESLint 9.
_(TBH, I only installed v9 because of #17772)_

```
npm ls
***@1.0.0 /home/***
├── @eslint/js@9.0.0-rc.0
├── @stylistic/eslint-plugin-js@1.7.0
├── eslint@9.0.0-rc.0
├── htmlhint@1.1.4
├── postcss-lit@1.1.1
├── stylelint-config-standard@36.0.0
└── stylelint@16.3.1
```

**Comment by aladdin-add** [comment]
try `npm install --legacy-peer-deps`. (if using npm>=7)

**Comment by erosman** [comment]
The initial issue is sorted with eslint@9.0.0

```
npm ls
***@1.0.0 /home/***
├── @eslint/js@9.0.0
├── @stylistic/eslint-plugin-js@1.7.0
├── eslint@9.0.0
├── htmlhint@1.1.4
├── postcss-lit@1.1.1
├── stylelint-config-standard@36.0.0
└── stylelint@16.3.1
```


I was wondering why there are multiple versions of many modules e.g.

```
node_modules/espree/package.json -> "version": "9.6.1",
node_modules/eslint/node_modules/espree/package.json -> "version": "10.0.1",

node_modules/eslint-visitor-keys/package.json -> "version": "3.4.3",
node_modules/eslint/node_modules/eslint-visitor-keys/package.json -> "version": "4.0.0",

node_modules/@eslint-community/
node_modules/eslint/node_modules/@eslint-community/
```

Is that how it should be?
If not, how can the extra ones be removed?

Also ....

```
npm why -r @eslint/js
@eslint/js@9.0.0 dev
node_modules/@eslint/js
  dev @eslint/js@"^9.0.0" from the root project
  @eslint/js@"9.0.0" from eslint@9.0.0
  node_modules/eslint
    dev eslint@"^9.0.0" from the root project
    peer eslint@">=8.40.0" from @stylistic/eslint-plugin-js@1.7.0
    node_modules/@stylistic/eslint-plugin-js
      dev @stylistic/eslint-plugin-js@"^1.7.0" from the root project
    peer eslint@"^6.0.0 || ^7.0.0 || >=8.0.0" from @eslint-community/eslint-utils@4.4.0
    node_modules/eslint/node_modules/@eslint-community/eslint-utils
      @eslint-community/eslint-utils@"^4.2.0" from eslint@9.0.0

```


The last one is referencing
`node_modules/eslint/node_modules/@eslint-community/eslint-utils` 
but there is no `eslint-utils` in
`node_modules/@eslint-community/regexpp`


Many nested `/node_modules/` ...
**node_modules**/eslint/**node_modules**/@eslint-community/eslint-utils/**node_modules**/eslint-visitor-keys/package.json


For example, how are these placed in 2 different locations?
```
node_modules/eslint/node_modules/@eslint-community/eslint-utils
node_modules/@eslint-community/regexpp
```

https://github.com/eslint/eslint/blob/5c353215e05818e17e83192acbb4d3730c716afa/lib/rules/utils/regular-expressions.js#L9

https://github.com/eslint/eslint/blob/5c353215e05818e17e83192acbb4d3730c716afa/lib/rules/no-obj-calls.js#L12
